src/java.base/share/classes/java/lang/reflect/code/parser/impl/DescParser.java

```diff
@@ -85,16 +88,42 @@ public static RecordTypeRef parseRecordTypeRef(String desc) {
     }
 
     public static TypeDefinition parseTypeDefinition(Lexer l) {
+        boolean isTypeVar = false;
+        if (l.token().kind == TokenKind.HASH) {
+            l.accept(TokenKind.HASH);
+            isTypeVar = true;
+        }
         // Type
-        Tokens.Token t = l.accept(Tokens.TokenKind.IDENTIFIER);
+        Tokens.Token t = l.accept(TokenKind.IDENTIFIER,
+                TokenKind.PLUS, TokenKind.SUB);
         StringBuilder identifier = new StringBuilder();
-        identifier.append(t.name());
+        identifier.append(t.kind == TokenKind.IDENTIFIER ? t.name() : t.kind.name);
         while (l.acceptIf(Tokens.TokenKind.DOT)) {
             identifier.append(Tokens.TokenKind.DOT.name);
             t = l.accept(Tokens.TokenKind.IDENTIFIER);
             identifier.append(t.name());
         }
 
+        if (l.token().kind == TokenKind.COLCOL && isTypeVar) {
```

Comment:
If we see `#Foo::Bar` we might be seeing two things:
* a type-variable `Bar` in class `Foo`
* a method type-variable in some method `Foo.Bar`

So, we need to disambiguate based on what follows. E.g. if after `Bar` we see `(`, then we know we're in the method case.